File: feb05/perl/sub_encode.pl
#!/usr/bin/perl

use strict;
use CGI qw/:standard/;

print header(-type=>'text/html',-expires=>'+1d');
print "<html><head><title>Substitution Cipher Results</title></head>\n\n";

print "<center><h2>Substitution Cipher Results</h2>";
print "<body bgcolor=\"#FF8844\">";
print "<table width=\"600\" border=\"1\" cellspacing=\"0\" cellpadding=\"2\"><tr><td>";
my $plaintext=param('plaintext');
my $cipher=param('cipher');
print "The original text was: ".$plaintext;
print "<br><br>\n\n The cipher was: ".$cipher;
if(length($cipher)!=26) {print "<br><br>\n\nThis cipher does not have 26 characters.";}
elsif ($cipher!~/[a-zA-Z]{26}/) {print "<br><br>\n\nYou have non-alphabetical characters in the cipher.";}
else {
  my $crypttext;
  $_=$plaintext;
  $cipher =~tr/A-Z/a-z/;
  eval "tr/a-z/$cipher/";
  $cipher =~tr/a-z/A-Z/;
  eval "tr/A-Z/$cipher/";
  print "<br><br>\n\n The translated text is: ".$_;
  print "</td></tr></table></center>";
  print "</body></html>";
}
1;

File: feb05/sub_encode.php
<html><head><title>Substitution Cipher Encoding</title></head>
<body bgcolor="#FF8844">
<script language="javascript">//<!--

function validate() {
  if(document.form1.cipher.value.length!=26) {alert('Invalid cipher length.'); return false;}
return true;
}
function randomCipher() {
 /* The algorithm we use will count from 25 to 0.  For each number n, it selects a character
    from 1 to n and switches that character with the nth, then decrements n.  Each character
    has a 1/26 chance of being in each spot in this way */
 string="abcdefghijklmnopqrstuvwxyz";
 for(i=25;i>0;i--) {
    j=Math.floor(Math.random()*(i+1));
    ic=string.charAt(i);
    jc=string.charAt(j);
    string=string.replace(ic,"%");
    string=string.replace(jc,ic);
    string=string.replace("%",jc);
 }
 document.form1.cipher.value=string;
}
//-->
</script>
<center><h2>Substitution Cipher</h2>
<table width="600" border="1" cellspacing="0" cellpadding="2"><tr><td>
Please enter the cipher as a string of 26 characters, none of which are repeated.  If a character is repeated, both of the corresponding plaintext characters will be translated to that value.
<form method="POST" name="form1" action="perl/sub_encode.pl">
<table><tr><td>Plain Text</td><td><textarea name="plaintext" rows="5" cols="50"></textarea></td></tr>
<tr><td>Cipher</td><td><input name="cipher" size="26" maxlength="26"> <input type="button" value="Generate Random" onClick="randomCipher();"></td></tr>
</table>
<input type="submit" value="Encode" onClick="return validate();">
</form>
</td></tr></table></center>
</body></html>